### ISSUE TITLE
Custom NER Model not giving correct results after saving

### ISSUE THREAD

**Issue by msolanki-cpu** [HassanSherwani/Custom_NER#1]
I am using the following standard code to train a dataset of resumes, testing another dataset and then saving it to disk.
Everything works fine till here and I am getting correctly predicted values to an accuracy levels of 99.6%.

However, when I am using the saved model to predict values in any file being passed to the model, the values are totally different.
I have searched this issue and tried every possible thing recommended already on github and none of them seem to have worked  for me. Hence, I want to request to open this issue again.

-------------------------------------------------------------------------------------------------------------
-------------------------------------------------------------------------------------------------------------
Environment Details -
OS - Windows 10 Enterprise
Spacy Version - 2.2.3
Python Version - 3.8.1

I am training a dataset of 100 resumes with 10 iterations at the moment with following steps -
1) Convert resume PDF to JSON
2) Pass JSON as train_Data

I am currently dealing with 4 entities only -
1) Companies Worked At
2) College
3) Degree
4) Location

Here is the code -

import json
import random
import logging
from sklearn.metrics import classification_report
from sklearn.metrics import precision_recall_fscore_support
from spacy.gold import GoldParse
from spacy.scorer import Scorer
from sklearn.metrics import accuracy_score
import re
import spacy
from numpy.ma.core import get_data
from bin.wiki_entity_linking.wikipedia_processor import _get_gold_parse

def trim_entity_spans(data: list) -> list:
    invalid_span_tokens = re.compile(r'\s')
    cleaned_data = []
    for text, annotations in data:
        entities = annotations['entities']
        valid_entities = []
        for start, end, label in entities:
            valid_start = start
            valid_end = end
            while valid_start < len(text) and invalid_span_tokens.match(
                    text[valid_start]):
                valid_start += 1
            while valid_end > 1 and invalid_span_tokens.match(
                    text[valid_end - 1]):
                valid_end -= 1
            valid_entities.append([valid_start, valid_end, label])
        cleaned_data.append([text, {'entities': valid_entities}])
    return cleaned_data

def convert_dataturks_to_spacy(dataturks_JSON_FilePath):
    try:
        training_data = []
        lines=[]
        with open(dataturks_JSON_FilePath, encoding="utf-8") as f:
            lines = f.readlines()            
        for line in lines:
            data = json.loads(line)
            text = data['content']
            entities = []
            data_annotations = data['annotation']
            if data_annotations is not None:
                for annotation in data_annotations:
                    #only a single point in text annotation.
                    point = annotation['points'][0]
                    labels = annotation['label']
                    # handle both list of labels or a single label.
                    if not isinstance(labels, list):
                        labels = [labels]
                    for label in labels:
                        point_start = point['start']
                        point_end = point['end']
                        point_text = point['text']

                        lstrip_diff = len(point_text) - len(point_text.lstrip())
                        rstrip_diff = len(point_text) - len(point_text.rstrip())
                        if lstrip_diff != 0:
                            point_start = point_start + lstrip_diff
                        if rstrip_diff != 0:
                            point_end = point_end - rstrip_diff
                        entities.append((point_start, point_end + 1 , label))
            training_data.append((text, {"entities" : entities}))           
            
        return training_data
    except Exception as e:
        logging.exception("Unable to process " + dataturks_JSON_FilePath + "\n" + "error = " + str(e))
        return None

def train_spacy():
    TRAIN_DATA = convert_dataturks_to_spacy("C:\\Users\\msolanki\\Downloads\\Entity-Recognition-In-Resumes-SpaCy-master\\traindata.json")
    TRAIN_DATA = trim_entity_spans(TRAIN_DATA)
    

    my_json_string = json.dumps(TRAIN_DATA)
    with open('data.json', 'w', encoding='utf-8') as f:
        json.dump(my_json_string, f, ensure_ascii=False, indent=4)
    print("train_data", type(TRAIN_DATA))
    nlp = spacy.blank('en')
    if 'ner' not in nlp.pipe_names:
        ner = nlp.create_pipe('ner')
        nlp.add_pipe(ner, last=True)
    for _, annotations in TRAIN_DATA:
         for ent in annotations.get('entities'):
            ner.add_label(ent[2])
    other_pipes = [pipe for pipe in nlp.pipe_names if pipe != 'ner']
    with nlp.disable_pipes(*other_pipes):  # only train NER
        optimizer = nlp.begin_training()
        for itn in range(10):
            print("Statring iteration " + str(itn))
            random.shuffle(TRAIN_DATA)
            losses = {}
            for text, annotations in TRAIN_DATA:
                nlp.update(
                [text],  # batch of texts
                [annotations],  # batch of annotations
                drop=0.2,  # dropout - make it harder to memorise data
                sgd=optimizer,  # callable to update weights
                losses=losses)
            print(losses)    
            

    nlp.to_disk("C:\\Users\\msolanki\\Downloads\\Entity-Recognition-In-Resumes-SpaCy-master\\spacymodel")
    #nlp.from_disk("C:\\Users\\akjain\\Downloads\\Entity-Recognition-In-Resumes-SpaCy-master\\spacymodel")
    examples = convert_dataturks_to_spacy("C:\\Users\\msolanki\\Downloads\\Entity-Recognition-In-Resumes-SpaCy-master\\testdata.json")
    examples = trim_entity_spans(examples)
    tp=0
    tr=0
    tf=0
    ta=0
    c=0        
    for text,annot in examples:
        f=open("resume"+str(c)+".txt","w", encoding ='utf-8')
        #nlp_trained = spacy.load('C:\\Users\\akjain\\Downloads\\Entity-Recognition-In-Resumes-SpaCy-master\\spacymodel')
        doc_to_test=nlp(text)
        d={}
        for ent in doc_to_test.ents:
            d[ent.label_]=[]
        for ent in doc_to_test.ents:
            d[ent.label_].append(ent.text)
        for i in set(d.keys()):
            f.write("\n\n")
            f.write(i +":"+"\n")
            for j in set(d[i]):
                f.write(j.replace('\n','')+"\n")
        d={}
        for ent in doc_to_test.ents:
            d[ent.label_]=[0,0,0,0,0,0]
        for ent in doc_to_test.ents:
            doc_gold_text= nlp.make_doc(text)
            gold = GoldParse(doc_gold_text, entities=annot.get("entities"))
            y_true = [ent.label_ if ent.label_ in x else 'Not '+ent.label_ for x in gold.ner]
            y_pred = [x.ent_type_ if x.ent_type_ ==ent.label_ else 'Not '+ent.label_ for x in doc_to_test]  
            if(d[ent.label_][0]==0):
                (p,r,f,s)= precision_recall_fscore_support(y_true,y_pred,average='weighted')
                a=accuracy_score(y_true,y_pred)
                d[ent.label_][0]=1
                d[ent.label_][1]+=p
                d[ent.label_][2]+=r
                d[ent.label_][3]+=f
                d[ent.label_][4]+=a
                d[ent.label_][5]+=1
        c+=1
    for i in d:
        print("\n For Entity "+i+"\n")
        print("Accuracy : "+str((d[i][4]/d[i][5])*100)+"%")
        print("Precision : "+str(d[i][1]/d[i][5]))
        print("Recall : "+str(d[i][2]/d[i][5]))
        print("F-score : "+str(d[i][3]/d[i][5]))
train_spacy()

--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Here is the code snippet I am using to load the saved model and pass a resume to it -

import spacy
import PyPDF2
#mypdf = open(r'C:\Users\msolanki\Downloads\All_Resumes\All_Resumes\G323381_Professional CV current.pdf', mode='rb')
#pdf_document = PyPDF2.PdfFileReader(mypdf)
#first_page = pdf_document.

nlp = spacy.load("C:\\Users\\msolanki\\Downloads\\Entity-Recognition-In-Resumes-SpaCy-master\\spacymodel") 
#text = first_page.extractText()	
with open(r'C:\Users\msolanki\Downloads\All_Resumes\All_Resumes\To be Uploaded\G278204_Robin Williams.pdf','rb') as pdf_file:
	read_pdf = PyPDF2.PdfFileReader(pdf_file)
	number_of_pages = read_pdf.getNumPages()
	
	for page_number in range(number_of_pages):
		page = read_pdf.getPage(page_number)
		page_content=page.extractText()
		doc=nlp(page_content)
		for ent in doc.ents:
			print(ent.text,ent.label_)

Please let me know if you will require more information from my end.

PS : I am unable to upload traindata and testdata(due to confidential reasons) as the resumes belong to the company I work at but I had tried the entire thing with sample dataset too and the same behavior exhibited. 

Will really appreciate any help at all here!
